pkg/operator/clustermemberremovalcontroller/clustermemberremovalcontroller.go

```diff
@@ -0,0 +1,259 @@
+	isMachineAPIFunctionalFn func() (bool, error)
+	masterMachineSelector    labels.Selector
+	masterNodeSelector       labels.Selector
+}
+
+// NewClusterMemberRemovalController removes an etcd member if the machine and a node for the etcd member is gone and the machine-api is active
+//
+// Note:
+//  since this controller needs to reconcile only master nodes and machine objects
+//  make sure nodeInformer and machineInformer contain only filtered data
+//  otherwise it might be expensive to react to every node update in larger installations
+func NewClusterMemberRemovalController(
+	etcdClient etcdcli.EtcdClient,
+	isMachineAPIFunctional func() (bool, error),
+	masterMachineSelector labels.Selector, masterNodeSelector labels.Selector,
+	kubeInformersForNamespaces operatorv1helpers.KubeInformersForNamespaces,
+	masterNodeInformer cache.SharedInformer,
+	masterMachineInformer cache.SharedInformer,
+	masterNodeLister corev1listers.NodeLister,
+	masterMachineLister machinelistersv1beta1.MachineLister,
+	networkInformer configv1informers.NetworkInformer,
+	eventRecorder events.Recorder,
+) factory.Controller {
+	c := &ClusterMemberRemovalController{
+		etcdClient:                        etcdClient,
+		isMachineAPIFunctionalFn:          isMachineAPIFunctional,
+		masterMachineSelector:             masterMachineSelector,
+		masterNodeSelector:                masterNodeSelector,
+		masterNodeLister:                  masterNodeLister,
+		masterMachineLister:               masterMachineLister,
+		networkLister:                     networkInformer.Lister(),
+		configMapListerForTargetNamespace: kubeInformersForNamespaces.InformersFor(operatorclient.TargetNamespace).Core().V1().ConfigMaps().Lister().ConfigMaps(operatorclient.TargetNamespace),
+	}
+	return factory.New().
+		WithSync(c.sync).
+		ResyncEvery(33*time.Minute). // make it slow since nodes are updated every few minutes
+		WithBareInformers(networkInformer.Informer()).
+		WithInformers(
+			masterNodeInformer,
+			masterMachineInformer,
+			kubeInformersForNamespaces.InformersFor(operatorclient.TargetNamespace).Core().V1().ConfigMaps().Informer(),
+		).ToController("ClusterMemberRemovalController", eventRecorder.WithComponentSuffix("cluster-member-removal-controller"))
+}
+
+func (c *ClusterMemberRemovalController) sync(ctx context.Context, _ factory.SyncContext) error {
+	// stop if the machine API is not functional
+	if isFunctional, err := c.isMachineAPIFunctionalFn(); err != nil {
+		return err
+	} else if !isFunctional {
+		return nil
+	}
+	etcdEndpointsConfigMap, err := c.configMapListerForTargetNamespace.Get("etcd-endpoints")
+	if err != nil {
+		return err // should not happen
+	}
+	if len(etcdEndpointsConfigMap.Data) == 0 {
+		return nil
+	}
+	for _, potentialMemberToRemoveIP := range etcdEndpointsConfigMap.Data {
+		memberLocator := potentialMemberToRemoveIP
+		machine, err := c.getMachineForMember(potentialMemberToRemoveIP)
+		if err != nil && err != errNotFound {
```

Comment:
if this were a real client, it would be

```go
switch{
case apierrors.IsNotFound(err):
   // this member is eligible for removal
case err != nil:
  return return fmt.Errorf("unable to get a machine for member: %v, err: %w", memberLocator, err)
default:
  klog.V(4).Infof("cannot remove member: %v because a machine resource still exists (%v/%v)", memberLocator, machine.Name, machine.UID)
			continue
}
```

Reply:
we need to be very careful here, technically we might get an `apierrors.IsNotFound` from a lister (`masterNodeLister`, `networkLister` or `masterMachineLister`) which would cause to remove a machine falsely, right?

Reply:
Good question. In general, many controllers solely rely on caches being synced. If we had a bug many of them wouldn't work properly. Although technically it is possible I don't think it can happen in practice. Mainly because caches must get a consistent snapshot of data before they are considered synchronized for the first time.

In addition to that even if the cache was stale we should be okay. The worst-case scenario would be that we wouldn't delete a member from the cluster. The other controllers must take a scenario like that into account anyway and wait until it happens.

To be honest I would prefer to return a `boolean` instead of a sentinel error.